modules/placement-driver/src/main/java/org/apache/ignite/internal/raft/client/TopologyAwareRaftGroupService.java

```diff
@@ -0,0 +1,443 @@
+                                                        serverEventHandler.onLeaderElected(appearedNode, leaderWithTerm.term());
+                                                    }
+                                                }, executor);
+                                    }
+
+                                    return CompletableFuture.completedFuture(null);
+                                }, executor);
+                    }
+                }
+            }
+        });
+    }
+
+    /**
+     * Starts an instance of topology aware RAFT client.
+     *
+     * @param groupId Replication group id.
+     * @param cluster Cluster service.
+     * @param factory Message factory.
+     * @param raftConfiguration RAFT configuration.
+     * @param configuration Group configuration.
+     * @param getLeader True to get the group's leader upon service creation.
+     * @param executor RPC executor.
+     * @param logicalTopologyService Logical topology service.
+     * @return Future to create a raft client.
+     */
+    public static CompletableFuture<RaftGroupService> start(
+            ReplicationGroupId groupId,
+            ClusterService cluster,
+            RaftMessagesFactory factory,
+            RaftConfiguration raftConfiguration,
+            PeersAndLearners configuration,
+            boolean getLeader,
+            ScheduledExecutorService executor,
+            LogicalTopologyService logicalTopologyService
+    ) {
+        return RaftGroupServiceImpl.start(groupId, cluster, factory, raftConfiguration, configuration, getLeader, executor)
+                .thenApply(raftGroupService -> new TopologyAwareRaftGroupService(cluster, factory, executor, raftConfiguration,
+                        raftGroupService, logicalTopologyService));
+    }
+
+    /**
+     * Sends a subscribe message to a specific node of the cluster.
+     *
+     * @param node Node.
+     * @param msg  Subscribe message.
+     * @return A future that will complete when the message is sent.
+     */
+    private CompletableFuture<Boolean> sendSubscribeMessage(ClusterNode node, SubscriptionLeaderChangeRequest msg) {
+        var msgSendFut = new CompletableFuture<Boolean>();
+
+        sendWithRetry(node, msg, msgSendFut);
+
+        return msgSendFut;
+    }
+
+    /**
+     * Tries to send a subscribe message until the node leaves of the cluster.
+     *
+     * @param node       Node.
+     * @param msg        Subscribe message to send.
+     * @param msgSendFut Future that will completed when the message is send or an issue prevent that.
+     */
+    private void sendWithRetry(ClusterNode node, SubscriptionLeaderChangeRequest msg, CompletableFuture<Boolean> msgSendFut) {
+        clusterService.messagingService().invoke(node, msg, raftConfiguration.responseTimeout().value()).whenCompleteAsync((unused, th) -> {
+            if (th != null) {
+                if (recoverable(th)) {
```

Comment:
seems that this throwable should be added as cause exception if the future is completed with another one.

Reply:
I think the loop will be internally in network service layer in the future. They handle it more correctly, but currently I think the node which is catching an exception due to cannot get a topology has no cause except the particular exception.
Moreover, we do not an exception when we retry send a message and node still available in logical topology. Why do we have to write an exception into cause if we cannot get a logical topology?.

Reply:
maybe suppressed then?